File: quadrature.py
from math import sin, e

def simpsons_rule(f, a, b):
    c = (a+b) / 2.000
    h3 = abs(b-a) / 6.0
    print("S(%0.10f, %0.10f) = %0.10f" % (a, b, h3*(f(a) + 4.0*f(c) + f(b))))
    return h3*(f(a) + 4.0*f(c) + f(b))

def recursive_asr(f, a, b, eps, whole, l):
    "Recursive implementation of adaptive Simpson's rule."
    print("level = %d" % l)
    c = (a+b) / 2.0
    left = simpsons_rule(f,a,c)
    right = simpsons_rule(f,c,b)
    if abs(left + right - whole) <= 15*eps:
        return left + right + (left + right - whole)/15.0
    return recursive_asr(f, a, c, eps/2.0, left, l+1) + recursive_asr(f, c, b, eps/2.0, right, l+1)

def adaptive_simpsons_rule(f, a, b, eps):
    "Calculate integral of f from a to b with max error of eps."
    return recursive_asr(f, a, b, eps, simpsons_rule(f, a, b), 0)


def main():
    print(adaptive_simpsons_rule(lambda x: (e**(2*x))*sin(3*x), 1, 3, 1e-5))

main()

File: quadrature.m
function retval = quadrature(f, a, b, t, whole)
    c = (a+b)./2;
    left = ((c-a)./3) * (f(a) + 4*f(a + (c-a)./2) + f(c));
    right = ((b-c)./3) * (f(c) + 4*f(c + (b-c)./2) + f(b));
    if (abs(left + right - whole) <= 15*t)
        Q = left + right + (left + right - whole)/15;
    else
        Q = quadrature(f, a, c, t/2, left) + quadrature(f, c, b, t/2, right);
    endif

    retval = Q;
    %l -= 1;
    %Q = ((b-a)./3) * (f(a) + 4*f(a + (b-a)./2) + f(b));
    % printf("level = %d\nS(%.10f, %.10f) = %.10f\n", l, a, b, Q);
    %[v, ier, nfun, err] = quad(f, a, b, t);
    %eps = abs(Q - v);
    ## if (eps > t & l > 0)
    ##     m = (a+b)./2;
    ##     Q = quadrature(f, a, m, t, l) + quadrature(f, m, b, t, l);
    ## endif
    ## retval = Q;
endfunction
